src/main/java/com/squareup/javapoet/CodeWriter.java

```diff
@@ -203,6 +206,10 @@ public void emitTypeVariables(List<TypeVariableName> typeVariables) throws IOExc
     emit(">");
   }
 
+  public void popTypeVariables(List<TypeVariableName> typeVariables) throws IOException {
+    typeVariables.forEach(typeVariable -> currentTypeVariables.remove(typeVariable.name));
```

Comment:
Won't this cause issues with type variable shadowing? (or will the compiler reject it anyway?)

Even without shadowing, how about type variables of the same name in inner elements? (i.e. a static method or an inner class)

E.g. when generating
```java
class MyClass<T> {
  static <T> newInstance(T foo) { return new MyClass<T>(foo); }

  private final com.squareup.tacos.T<T> foo;

  private MyClass(com.squareup.tacos.T<T> foo) { this.foo = foo; }

  public com.squareup.tacos.T<T> getFoo() { return foo; }
}
```
won't this remove `T` from the set when exiting the `newInstance` method?

A more realistic example being:
```java
class MyClass<T> {
  static class Inner<T> {}

  private com.squareup.tacos.T<T> foo;
}
```
Will the `T` type variable still be in scope when exiting `Inner`?

Reply:
You're right. I don't think there's a way to then reference the "outer" `T`, but we do still need to have something closer to a multiset

Reply:
@JakeWharton @swankjesse do you have an alternative approach to a multiset? I know we don't depend on Guava here, so I was about to implement a Map<String, Integer>, but that felt gross, so I was thinking of creating a private Multiset type to wrap it. Thoughts?

Reply:
I went with the basic multiset approach, but happy to switch to another alternative.